<h4>Classifying epileptic EEG signals</h4>

Epilepsy is a medical condition that affects around 50 million people globally according to the World Health Organization as of February 2024 (https://www.who.int/news-room/fact-sheets/detail/epilepsy). Here we will use There is evidence that suggests that the brain signal becomes alterated sometime before the seizure by the brain Here we will use a small sample of brain recordings to 

Handa, Palak, Monika Mathur, and Nidhi Goel. "Open and free EEG datasets for epilepsy diagnosis." arXiv preprint arXiv:2108.01030 (2021).

In [77]:
import pathlib

dataset_path = pathlib.Path('EEG-Epilepsy-Datasets')

with open(dataset_path / 'Datasets Information.txt', 'r') as description:
    text = description.read()
print(text)

Datasets:
These datasets are exemplary segmented EEG time series recordings of ten epilepsy patients collected from Neurology & Sleep Centre, Hauz Khas, New Delhi. The data was acquired at sampling rate of 200 Hz using Grass Telefactor Comet AS40 Amplification System. During the acquisition, gold plated scalp EEG electrodes were placed according to 10-20 electrode placement system. The signals were filtered between 0.5 to 70 Hz and then segmented into pre-ictal, interictal and ictal stages. Each downloadable folder contains fifty MAT-files of EEG time series signals. The name of the folder corresponds to the epileptic seizure stage. Each MAT-file consists of 1024 samples of one EEG time series data lasting for a duration of 5.12 seconds.     

Note: The datasets are part of an ongoing research work in collaboration of Neurocomputing Laboratory at IIT Delhi and Neurology & Sleep Centre, Hauz Khas. Kindly cite this link and following manuscript(s) when you publish results based the EEG s

In [22]:
from collections import defaultdict

import scipy.io
import torch
import torch.nn as nn
from torch import Tensor
from torch.utils.data import Dataset
from torch.utils.data import DataLoader

**Exercise 1** - Load the eeg data from the EEG-Epilepsy-Datasets folder. The data is organized in three subfolders: the first one contains EEG recordings from the ictal phase, the second one has EEG from the preictal phase and the last one has EEG from interictal (normal) brain state. Use the library pathlib to iterate over the files and load the EEGs

In [23]:
dataset_path = pathlib.Path('EEG-Epilepsy-Datasets')
eeg_signal_types = ['ictal', 'interictal', 'preictal']
eeg = defaultdict(list)
for eeg_type in eeg_signal_types:
    eeg_type_path = dataset_path / eeg_type
    for file in eeg_type_path.iterdir():
        eeg[eeg_type].append(scipy.io.loadmat(file))

We now have a dictionary with three keys: 'ictal', 'interictal' and 'preictal'. Each of the keys has an associated list value with all the recordings for that particular EEG state. Let's see how many recordings of each type we have:

In [79]:
print(len(eeg['ictal']))
print(len(eeg['interictal']))
print(len(eeg['preictal']))

50
50
50


Each recording has 1024 EEG datapoints that result from 5.12 seconds of measurement at a sampling rate of 200Hz. In addition to the EEG data, each .mat file that we loaded has a description of the recording:

In [81]:
eeg['ictal'][0]

{'__header__': b'MATLAB 5.0 MAT-file, Platform: PCWIN64, Created on: Sat Aug 06 14:02:24 2016',
 '__version__': '1.0',
 '__globals__': [],
 'ictal': array([[-38],
        [-26],
        [ -9],
        ...,
        [131],
        [136],
        [134]], dtype=int16)}

**Exercise 2** From each The ictal state  The elements of the list should look as follows:

In [ ]:
class EEGDataset(Dataset):
    def __init__(self, data: List[Tuple[Tensor, Tensor]]):
        self.data = data

    def __len__(self):
        return len(self.data)

    def __getitem__(self, idx):
        return self.data[idx]

**Exercise 3** - Create a class called `NeuralNetwork` that inherits from `nn.Module` which will represent our model. In the `forward` method, 

In [21]:
os.path.join('EEG-Epilepsy-Datasets', 'ictal')

'EEG-Epilepsy-Datasets/ictal'

In [ ]:
class NeuralNetwork(nn.Module):
    def __init__(self):
        super().__init__(self)
        # your code goes here

    def forward(self):
        # your code goes here
        pass

**Exercise 5** - Define the training loop. 

<h4>References</h4>
[1] Handa, Palak, Monika Mathur, and Nidhi Goel. "Open and free EEG datasets for epilepsy diagnosis." arXiv preprint arXiv:2108.01030 (2021).

[2] Swami, Piyush, et al. "EEG epilepsy datasets." DOI: https://doi.org/10.13140/RG 2.14280.32006 (2016).